File: Golang/ImageRegistration/image-registration.go
package main

import (
	"fmt"
	"image"
	"image/color"
	"os"

	"gocv.io/x/gocv"
)

var redColor = color.RGBA{R: 255}
var bx, by, ax, ay int
var segment int = 5

const minPos = 9.9e+25

func main() {
	// declare window
	// window1 := gocv.NewWindow("image 1")
	// window2 := gocv.NewWindow("image 2")
	// windowResult := gocv.NewWindow("result")

	imgPath1 := "../../data/imageRegistration-1.jpg"
	imgPath2 := "../../data/imageRegistration-2.jpg"
	img1 := gocv.IMRead(imgPath1, gocv.IMReadColor)
	img2 := gocv.IMRead(imgPath2, gocv.IMReadColor)
	if img1.Empty() {
		fmt.Printf("Could not read image %s\n", imgPath1)
		os.Exit(1)
	}
	if img2.Empty() {
		fmt.Printf("Could not read image %s\n", imgPath2)
		os.Exit(1)
	}

	col := img1.Cols() / segment
	row := img1.Rows() / segment

	for i := 0; i < segment; i++ {
		for j := 0; j < segment; j++ {
			// create rectangle
			rect := image.Rect(i*col, j*row, (i+1)*col, (j+1)*row)

			// crop image
			tmp := img1.Region(rect)

			result := img1.Clone()

			// match template
			mask := gocv.NewMat()
			gocv.MatchTemplate(img2, tmp, &result, gocv.TmSqdiff, mask)
			mask.Close()
			minVal, _, minLoc, _ := gocv.MinMaxLoc(result)

			if minVal < minPos {
				bx = minLoc.X
				by = minLoc.Y
				ax = i * col
				ay = j * row
			}
		}
	}

	gocv.Rectangle(&img1, image.Rect(ax, ay, ax+col, ay+row), redColor, 1)
	gocv.Rectangle(&img2, image.Rect(bx, by, bx+col, by+row), redColor, 1)
	result := gocv.NewMatWithSize(ay+img2.Rows()-by, ax+img2.Cols()-bx, gocv.MatTypeCV8UC3)

	// attach image1 to result
	roi := image.Rectangle{
		Min: image.Point{X: 0, Y: 0},
		Max: image.Point{X: img1.Size()[1], Y: img1.Size()[0]},
	}
	resultRoi := result.Region(roi)
	gocv.Resize(img1, &resultRoi, roi.Size(), 0, 0, gocv.InterpolationLinear)

	// attach image2 to result
	roi = image.Rectangle{
		Min: image.Point{X: ax - bx, Y: ay - by},
		Max: image.Point{X: result.Size()[1], Y: result.Size()[0]},
	}
	defer result.Close()
	resultRoi = result.Region(roi)
	defer resultRoi.Close()
	gocv.Resize(img2, &resultRoi, roi.Size(), 0, 0, gocv.InterpolationLinear)

	gocv.IMWrite("../../result/imageRegistrationResult.png", result)

	// window1.IMShow(img1)
	// window2.IMShow(img2)
	// windowResult.IMShow(result)
	// windowResult.WaitKey(0)
}

File: C++/ImageRegistration.cpp
#include "opencv4/opencv2/imgproc.hpp"
#include "opencv4/opencv2/imgcodecs.hpp"
#include "opencv4/opencv2/highgui.hpp"
#include "iostream"

using namespace cv;
using namespace std;


int main(int argc, char** argv)
{
	namedWindow("gambar 1", WINDOW_AUTOSIZE);
	namedWindow("gambar 2", WINDOW_AUTOSIZE);
	namedWindow("gabungan", WINDOW_AUTOSIZE);
	Mat hasil;
	double min, max;
	Point minL, maxL, B;
	int ax, ay;
	Mat img1 = imread(("../data/imageRegistration-1.jpg"), IMREAD_COLOR);
	Mat img2 = imread(("../data/imageRegistration-2.jpg"), IMREAD_COLOR);
	int jml_kolom = img1.cols / 4;
	int jml_baris = img1.rows / 4;
	for (int i = 0; i < 4; i++) {
		for (int j = 0; j < 4; j++) {
			Rect rec(i*jml_kolom, j*jml_baris, jml_kolom, jml_baris);
			Mat tmp = img1(rec).clone();
			matchTemplate(img2, tmp, hasil, TM_SQDIFF);
			minMaxLoc(hasil, &min, &max, &minL, &maxL);
			
			if (min < 2.4e+06) {
				rectangle(img1, rec, Scalar(0, 0, 255), 1, 8);
				rectangle(img2, Rect(minL.x, minL.y, jml_kolom, jml_baris), Scalar(0, 0, 255), 1, 8);
				ax = i * jml_kolom;
				ay = j * jml_baris;
				B.x = minL.x;
				B.y = minL.y;
				cout << "min:" << min << ",	max:" << max << "	(" << ax << "," << ay << ")" << endl;
			}
			imshow("gambar 1", img1);
			imshow("gambar 2", img2);
			//waitKey(0);
		}
	}
	//cout << ax << "-" << ay << "-" << img2.cols << "-" << img2.rows << "-" << minL.x << "-" << minL.y << endl;
	Mat gabungan = Mat::zeros(Size(ax + img2.cols - B.x, ay + img2.rows - B.y),CV_8UC3);
	//cout << "\n" << gabungan.cols << "-" << gabungan.rows << endl;
	img1.copyTo(gabungan(Rect(0, 0, img1.cols, img1.rows)));
	img2.copyTo(gabungan(Rect(ax - B.x, ay - B.y, img2.cols, img2.rows)));
	imshow("gabungan", gabungan);
	waitKey(0);
}